2 Bit Embedding Conversion Operator support.

radkris-git opened the pull request:
Summary:
2 Bit Embedding weight conversion operation is quite similar to
4 bit embedding weight conversion.

The diff contains both the
1. 2bit packing op `embedding_bag_2bit_prepack`.
2. 2bit unpacking op `embedding_bag_2bit_unpack`.

Comments about the op are inline with the op definition.

Test Plan: buck test caffe2/test:quantization -- test_embedding_bag_2bit_unpack

Differential Revision: D23143262

supriyar commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -133,9 +138,32 @@ Tensor qembeddingbag_4bit_prepack(const Tensor& weight) {
+// 15.
+// To later de-quantize values, the scale (range / 15) and zero_point
+// are stored alongside the data. More precisely, each row first has quantized
+// values, and then 2-byte fp16 scale and 2-byte zero_offset.)
```
nit: remove extra `)`

supriyar commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -133,9 +138,32 @@ Tensor qembeddingbag_4bit_prepack(const Tensor& weight) {
 }
 
+// Applies 4-bit row-wise quantization by determining the range
+// (maximum - minimum) and offset (minimum value) of each row in the input
```
call it bias instead of offset? We don't use offset anywhere in the code

supriyar commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -133,9 +138,32 @@ Tensor qembeddingbag_4bit_prepack(const Tensor& weight) {
+// To later de-quantize values, the scale (range / 3) and zero_point
+// are stored alongside the data. More precisely, each row first has quantized
+// values, and then 2-byte fp16 scale and 2-byte zero_offset.)
+Tensor qembeddingbag_2bit_prepack(const Tensor& weight) {
```
Might be worthwhile adding a comment (or TODO) here about the lookup operator not being added yet.

supriyar reviewed (approved):
Looks good, left some minor comments.

Commit:
2 Bit Embedding Conversion Operator support. (#43077)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/43077

2 Bit Embedding weight conversion operation is quite similar to
4 bit embedding weight conversion.

The diff contains both the
1. 2bit packing op `embedding_bag_2bit_prepack`.
2. 2bit unpacking op `embedding_bag_2bit_unpack`.

Comments about the op are inline with the op definition.

Test Plan: buck test caffe2/test:quantization -- test_embedding_bag_2bit_unpack

Reviewed By: supriyar

Differential Revision: D23143262

fbshipit-source-id: 97cd6da969abafb5d302b3150ba5349117f22b8a
